### ISSUE TITLE
@connection seems to be partially supported... doesn't work with belongsTo

### ISSUE THREAD

**Issue by rodrigoelp** [aws-amplify/amplify-ios#277]
This ticket is an escalation of #258 and  #254 (and I have closed #258 as I can see various problems with the way `@connection` behaves).

The content of this ticket might be addressed with documentation although I believe it is missing functionality from the amplify framework (or possibly aws-sdk-ios).

If you want to skip all the blab at the beginning, just to TL;DR;

## Description

The documentation tells us a **one to many** relationship can be specified as:

### Option 1: implicit connection definition (or autogenerated connection definition)
```graphql
type Parent
  @model  {
  id: ID!
  name: String!
  children: [Child] @connection
}

type Child
  @model {
  id: ID!
}
```
Meaning the relationship will be created specifying a connection as 
```graphql
input CreateParentChildInput {
  id: ID!
  _version: Int
  parentChildrenID: ID
}
```
And a companion swift code as:

```swift
public struct Parent: Model {
  public let id: String
  public var name: String
  public var children: List<Child>?

  public init(id: String = UUID().uuidString,
      name: String,
      children: List<Child>? = []) {
      self.id = id
      self.name = name
      self.children = children
  }
}

public struct Child: Model {
  public let id: String
  public var parentChildrenId: String?
  
  public init(id: String = UUID().uuidString,
      parentChildrenId: String? = nil) {
      self.id = id
      self.parentChildrenId = parentChildrenId
  }
}
```

To create these objects in swift we will need to:
```swift
func whatever() {
  let parent = Parent(id: UUID().uuidString, 
                            name: "Awesome adventures") // check issue #258 to understand why I do not provide the children here.

  Amplify.Api.mutate(of: parent, type: .create) { event in
   // if only event had map to allow us to map on success...
    switch event {
        case .completed(let result):
            switch result {
            case .success(let p):
                Amplify.Api.mutate(of: Child(parentChildrenId: p.id), type: .create) { event2 in
                    print(String(description: event2))
                }
            case .failure(let error):
                print("Got failed result with \(error.errorDescription)")
            }
        case .failed(let error):
            print("Got failed event with error \(error)")
        default:
            print("Should never happen")
        }
    }
  }
}
```

The **problem** is, this won't work. It fails with the following error:

```
failure(GraphQLResponseError<Child>: GraphQL service returned a successful response containing errors: [Amplify.GraphQLError(message: "Validation error of type FieldUndefined: Field \'parentChildrenId\' in type \'Child\' is undefined @ \'createChild/parentChildrenId\'", locations: Optional([Amplify.GraphQLError.Location(line: 4, column: 5)]), path: nil, extensions: nil)]
```

### Option 2: explicit connection definition.

This option follows every step as specified in the [documentation](https://aws-amplify.github.io/docs/cli-toolchain/graphql#connection) for **has many**.
```graphql
type Parent
  @model  {
  id: ID!
  name: String!
  children: [Child] @connection(keyName: "byParent", fields: ["id"])
}

type Child
  @model
  @key(name: "byParent", fields: ["parentId"]) {
  id: ID!
  parentId: ID!
}
```

Here you are forced to specify the id on the child element (which works fine), so I will not discuss this option as it works. But below is the code to save those instances in DymanoDB.

```swift
func whatever() {
  let parent = Parent(id: UUID().uuidString, 
                            name: "Awesome adventures") // check issue #258 to understand why I do not provide the children here.

  Amplify.Api.mutate(of: parent, type: .create) { event in
   // if only event had map to allow us to map on success...
    switch event {
        case .completed(let result):
            switch result {
            case .success(let p):
                // Instead of using the autogenerated field as in option 1, we use the specified one.
                Amplify.Api.mutate(of: Child(parentId: p.id), type: .create) { event2 in
                    print(String(description: event2))
                }
            case .failure(let error):
                print("Got failed result with \(error.errorDescription)")
            }
        case .failed(let error):
            print("Got failed event with error \(error)")
        default:
            print("Should never happen")
        }
    }
  }
}
```

### Option 3: explicit connection definition with belongs to connection (bi-directional connection).

This option is an extension of **Option 2** but it allows you to use the instance of your model instead of the key by itself... and here is the second oddity found in the framework. This option also breaks. Let's take a look at the type definition:
```graphql
type Parent
  @model  {
  id: ID!
  name: String!
  children: [Child] @connection(keyName: "byParent", fields: ["id"])
}

type Child
  @model
  @key(name: "byParent", fields: ["parentId"]) {
  id: ID!
  parentId: ID!
  # just adding this new line
  parent: Parent @connection(fields: ["parentId"])
}
```

The effect of this definition is, the property/attribute `var parentId: String` is replaced with the `var parent: Parent` in my autogenerated code as illustrated below:
```swift
public struct Parent: Model {
  public let id: String
  public var name: String
  public var children: List<Child>?

  public init(id: String = UUID().uuidString,
      name: String,
      children: List<Child>? = []) {
      self.id = id
      self.name = name
      self.children = children
  }
}

public struct Child: Model {
  public let id: String
  public var parentChildrenId: String?
  
  public init(id: String = UUID().uuidString,
      parent: Parent? = nil) {
      self.id = id
      self.parent = parent
  }
}
```

Our trusty function used to store data will be changed to:

```swift
func whatever() {
  let parent = Parent(id: UUID().uuidString, 
                            name: "Awesome adventures") // check issue #258 to understand why I do not provide the children here.

  Amplify.Api.mutate(of: parent, type: .create) { event in
   // if only event had map to allow us to map on success...
    switch event {
        case .completed(let result):
            switch result {
            case .success(let p):
                // Instead of using the autogenerated field as in option 1, we use the specified one.
                Amplify.Api.mutate(of: Child(parent: p), type: .create) { event2 in
                    print(String(description: event2))
                }
            case .failure(let error):
                print("Got failed result with \(error.errorDescription)")
            }
        case .failed(let error):
            print("Got failed event with error \(error)")
        default:
            print("Should never happen")
        }
    }
  }
}
```

Unfortunately, this option doesn't work **even though it is following step by step what the documentation indicates**. Here is the failure:

```
failure(GraphQLResponseError<Child>: GraphQL service returned a successful response containing errors: [Amplify.GraphQLError(message: "Variable \'input\' has coerced Null value for NonNull type \'ID!\'", locations: Optional([Amplify.GraphQLError.Location(line: 1, column: 31)]), path: nil, extensions: nil)]
```
I _assume_ the problem here is, the code associated with pulling out the default key (or the field associated with the instance) has not been pulled out and it provided `nil` instead, or it is following the incorrect path at the backend.

## TL;DR;
The definition with `@connection(fields: ["field"])` or `@connection` for one to many relationships does not seem to work. It only works with explicitly stated connections.

**Comment by amuresia** [comment]
+1 I was about to start tracking the exact same issue. I spent two days trying to understand what I did wrong although I was following the documentation step by step. This makes sdk version 3 unusable for APIs that are trying to mimic relational models. This is also visible in the DataStore service.

**Comment by rodrigoelp** [comment]
I don't know you @amuresia, but the `DataStore` is almost completely useless to us. It seems it is not even implemented... although to be honest, we decided 15 days ago to go raw GraphQL as we have hit so many limitations with amplify that we were doing more workarounds than anything else.

**Comment by amuresia** [comment]
@palpatim does this issue deserve the enhancement label or would it be better as a bug? The reason why I am asking is because there are several examples in the documentation, that talk about how to use amplify with complex models and none of them work.
https://aws-amplify.github.io/docs/ios/datastore#saving-relations

IMHO there is nothing new to be added, rather existing things to be fixed?

Also, may I be cheeky and ask for an ETA for the fix? Even if it's a vague one, it would be incredibly useful as it would allow me to plan other things around it.

Thank you and the team very much for your hard work on this!

**Comment by amuresia** [comment]
Hi, could we get an update on this please?

**Comment by amuresia** [comment]
Cold we get an update please?

**Comment by dtelaroli** [comment]
+1

**Comment by cabeaulac** [comment]
First tried DataStore, then Amplify API. I implemented Option 3 from the list above and couldn't figure out why the Parent was not being created in DynamoDB and the children are, and I'm getting a success response for all create mutations. Very tired of implementing work-arounds for all the bugs and having to read the small amount of unit tests to figure out how to use the API.

+1

**Comment by cabeaulac** [comment]
And I agree with @amuresia . This is not an enhancement. This is a bug, a very large bug.

**Comment by drochetti** [comment]
Hey @rodrigoelp @cabeaulac @amuresia @dtelaroli,

Thanks for you patience as I'm investigating this issue. There's a lot of info here and I'm trying to make sure I understand what's going on.

I started with the following schema, which I believe is the simplest representation of what was initially reported:

```graphql
type Parent @model {
  id: ID!
  name: String!
  children: [Child] @connection(name: "ParentChild")
}

type Child @model {
  id: ID!
  name: String!
  parent: Parent! @connection(name: "ParentChild")
}
```

With that schema I was able to successfully create a Parent and then a Child of that parent. I have shared my sample app with that schema in a personal repo for you to take a look: https://github.com/drochetti/AmplifyIssue277 (take a look at [ContentView.swift](https://github.com/drochetti/AmplifyIssue277/blob/master/Sources/ContentView.swift)) for the code that actually calls `Amplify.API`.

@cabeaulac you recently tried and experienced the same issue, right? Can you provide what version of Amplify are you using?

**Comment by dtelaroli** [comment]
I have solved the issue using a custom @key and referencing it with @connectino keyName property.

**Comment by drochetti** [comment]
Closing this issue as it seems to be resolved. If you're still running into the same issue and believe this was closed by mistake, feel free to reopen it.

**Comment by rodrigoelp** [comment]
> With that schema I was able to successfully create a Parent and then a Child of that parent. I have shared my sample app with that schema in a personal repo for you to take a look: https://github.com/drochetti/AmplifyIssue277 (take a look at [ContentView.swift](https://github.com/drochetti/AmplifyIssue277/blob/master/Sources/ContentView.swift)) for the code that actually calls `Amplify.API`.

That's good, that code looks fine to me.
